### PR TITLE
nested record updates

### PR DESCRIPTION
**Pull request by robin-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ci/da-ghc-lib/compile.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
   variables:
     ghc-lib-sha: '77fcc42eef2256ba9bb61a7e442a65bedb19a35d'
     base-sha: '9c787d4d24f2b515934c8503ee2bbd7cfac4da20'
-    patches: '323dc6fcb127f77c08423873efc0a088c071440a 833ca63be2ab14871874ccb6974921e8952802e9'
+    patches: '323dc6fcb127f77c08423873efc0a088c071440a 833ca63be2ab14871874ccb6974921e8952802e9' 'ce802bbf81e25f919381e10b0e589a59e7a59263'
```
Comment (cocreature) [inline ci/da-ghc-lib/compile.yml:14]:
```suggestion
    patches: 'ce802bbf81e25f919381e10b0e589a59e7a59263 833ca63be2ab14871874ccb6974921e8952802e9' ''
```
you shouldn’t add a new patch line. You need to replace the existing one.

File: ci/da-ghc-lib/compile.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ jobs:
   variables:
     ghc-lib-sha: '77fcc42eef2256ba9bb61a7e442a65bedb19a35d'
     base-sha: '9c787d4d24f2b515934c8503ee2bbd7cfac4da20'
-    patches: '323dc6fcb127f77c08423873efc0a088c071440a 833ca63be2ab14871874ccb6974921e8952802e9'
+    patches: 'ce802bbf81e25f919381e10b0e589a59e7a59263 833ca63be2ab14871874ccb6974921e8952802e9' ''
```
Comment (moritzkiefer-da) [inline ci/da-ghc-lib/compile.yml:14]:
```suggestion
    patches: 'ce802bbf81e25f919381e10b0e589a59e7a59263 833ca63be2ab14871874ccb6974921e8952802e9'
```

File: ci/da-ghc-lib/compile.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ jobs:
   variables:
     ghc-lib-sha: '77fcc42eef2256ba9bb61a7e442a65bedb19a35d'
     base-sha: '9c787d4d24f2b515934c8503ee2bbd7cfac4da20'
-    patches: '323dc6fcb127f77c08423873efc0a088c071440a 833ca63be2ab14871874ccb6974921e8952802e9'
+    patches: '323dc6fcb127f77c08423873efc0a088c071440a ce802bbf81e25f919381e10b0e589a59e7a59263'
```
Comment (cocreature) [inline ci/da-ghc-lib/compile.yml:14]:
```suggestion
    patches: 'ce802bbf81e25f919381e10b0e589a59e7a59263 833ca63be2ab14871874ccb6974921e8952802e9'
```
you replaced the wrong one. The first one was `da-master-8.8.1`. The second one is https://github.com/digital-asset/ghc/tree/da-unit-ids-8.8.1 which has our custom unit ids.

**Comment by robin-da** [comment]
changed the hash, but shouldn't matter though. this was already build against the new daml-lib-parser and fails with the expected error of not finding GHC.Prim.

**Comment by cocreature** [comment]
Interestingly if you look at the first failure, it is failing in the ghc-lib tests not our own code. So the cause seems to be completely unrelated to anything in the Daml repo.

**Comment by cocreature** [comment]
It looks like it got further now. It built ghc-lib and is now failing on a regular compile error in our code.

**Comment by cocreature** [comment]
Oh no, it’s failing on hlint so we should patch that.

File: compiler/damlc/tests/daml-test-files/EmptyWith.daml [hunk 4]
```diff
@@ -15,4 +13,4 @@ create = Foo with -- this should be fine
 match (Bar with) = 1 -- this should be fine
 
 update : Bar -> Bar
-update x = x with -- this should be an error (according to Haskell)
+update x = x with -- this should be be fine (since record dot updates)
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/EmptyWith.daml:16]: is this also accepted in GHC? I don’t really get why record dot updates change whether an empty record update is valid or not.

File: compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+-- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its affiliates.
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml:1]: Looks like a great set of tests, a few missing case:
1. Nested field puns
2. errors on typos in field names. In particular interested in locations there. Maybe better to test that in the shake tests, those already have a set of tests that test source locations.

**Review by cocreature — APPROVED** [review]
Looks great, thank you!

File: compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml (hunk 5 above)
Comment (robin-da) [reply compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml:1]: I think I already got nested field puns in the last line of that file. Adding an error location test in the shake tests.

Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml:1]:
```
y {text, inner.text = text, int, inner.int = int}
```
I only see non-nested field puns here. Am I looking at the wrong line?

Comment (robin-da) [reply compiler/damlc/tests/daml-test-files/RecordDotUpdates.daml:1]: Added the line `z = let Outer {..} = y in Outer {..}` to the tests.

File: ci/da-ghc-lib/compile.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ jobs:
   variables:
     ghc-lib-sha: '77fcc42eef2256ba9bb61a7e442a65bedb19a35d'
     base-sha: '9c787d4d24f2b515934c8503ee2bbd7cfac4da20'
-    patches: '323dc6fcb127f77c08423873efc0a088c071440a 833ca63be2ab14871874ccb6974921e8952802e9'
+    patches: '0c65ffa02eac181b89809bd3c88cf7b3af7fc38a 833ca63be2ab14871874ccb6974921e8952802e9'
```
[the remark this answers is not in the document]
Comment (robin-da) [reply ci/da-ghc-lib/compile.yml]: merged the ghc pr and updated the commit.

**Pull request merged**